returnn/tf/layers/basic.py

```diff
@@ -4812,16 +4830,31 @@ def get_out_data_from_opts(cls, n_out=NotSpecified, out_type=None, sources=(), *
     :param list[LayerBase] sources:
     :rtype: Data
     """
-    if n_out is NotSpecified and not out_type:
-      if not sources[0] or sources[0].output.undefined:
-        return Data.create_undefined(name="%s_output" % kwargs["name"])
-      out_type = sources[0].output.get_kwargs()
-      out_type["name"] = "%s_output" % kwargs["name"]
-      if out_type.get("sparse", False):
-        out_type["dim"] = 2  # True or False
-      out_type["dtype"] = "bool"
-      out_type["vocab"] = None
-    return super(CompareLayer, cls).get_out_data_from_opts(n_out=n_out, out_type=out_type, sources=sources, **kwargs)
+    out_type_ = {}
+    sources_ = [s for s in sources if s and not s.output.undefined]
+    if sources and not sources_:
+      if n_out is NotSpecified and not out_type:
+        from returnn.tf.network import CannotHandleUndefinedSourcesException
+        raise CannotHandleUndefinedSourcesException(
+          layer_name=kwargs["name"], layer_desc=dict(
+            n_out=n_out, out_type=out_type, sources=sources, **kwargs))
+    sources = sources_
+    if sources:
+      out_type_.update(Data.get_common_data([s.output for s in sources], warnings_out=log.v4).get_kwargs())
+    if n_out is not NotSpecified:
+      out_type_["dim"] = n_out
+    elif out_type.get("sparse", False):
```

Comment:
Sorry, this is wrong. It should be:
```suggestion
    elif out_type_.get("sparse", False):
```

Reply:
Please run some of the tests yourself. You would have seen that directly.